Allow built-in image paste handling to be disabled (#4874).

FlowIT-JIT opened the pull request:
This will allow 3rd party plugins to handle images being pasted.
Related issue: https://github.com/ckeditor/ckeditor4/issues/4874

## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [X] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [X] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4874](https://github.com/ckeditor/ckeditor4/issues/4874): Allow 3rd party plugins to handle images being pasted.
```

## What changes did you make?

Added a new configuration option to _clipboard_ plugin which allows for the built-in image paste handling to be disabled, so that other plugins can intercept clipboard data and handle images being pasted.

```
CKEDITOR.replace( 'editor', {
	// Allow 3rd party plugins to handle image pasting
	clipboard_handleImagePasting: false,
	extraPlugins: "customImagePasteHandlerPlugin"
} );
```

## Which issues does your PR resolve?

Closes #4874.

Comandeer commented on plugins/clipboard/plugin.js:
```diff
@@ -149,7 +149,7 @@
 			// custom MIME types in clipboard (#4612).
 			// Do it as the first step as the conversion is asynchronous and should hold all further paste processing.
-			if ( CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.plugins.clipboard.isFileApiSupported ) {
+			if ( ( CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.plugins.clipboard.isFileApiSupported ) && editor.config.clipboard_handleImagePasting !== false ) {
```
This condition became too long and therefore unreadable. Could you extract its first part to some variable (e.g. `isImagePastedSupported`)?

Additionally `editor.config.clipboard_handleImagePasting` alone should be enough to check if the plugin is supposed to handle image pasting.

FlowIT-JIT replied:
I just committet a change that shortens the code by extracting the first condition into an `isImagePasteSupported` variable as suggested. I'm not too crazy about the idea having only clipboard_handleImagePasting determine whether to enable image handling, since it will allow a developer to enable the feature on unsupported browsers.

Comandeer replied:
>  I'm not too crazy about the idea having only clipboard_handleImagePasting determine whether to enable image handling, since it will allow a developer to enable the feature on unsupported browsers.

Actually, I meant something different, but I see that my message could be ambiguous. I thought about shortening `editor.config.clipboard_handleImagePasting !== false` to `editor.config.clipboard_handleImagePasting`.

FlowIT-JIT replied:
Oh, okay 😅😊 Got it!

Comandeer commented on tests/plugins/clipboard/manual/pasteimagedisable.md:
```diff
@@ -0,0 +1,17 @@
+
+ 1. Paste an image into the editor.
+
+    Observe that the image is inserted and that no "Custom image paste handling" entry is written to "log output" on the page.
```
I'd split it into two tests: the one that will test the plugin with handling image pasting and one without it. It should be more readable than putting both of these checks inside one test. Could you split the test?

FlowIT-JIT replied:
I added unit tests to cover scenarios where image paste handling has been disabled using the new `clipboard_handleImagePasting` flag - see tests/plugins/clipboard/pasteimagedisabled.
Native image paste handling (when enabled) is already covered by tests/plugins/clipboard/pasteimage.

Do you still want manual test coverage, or should I just remove it again ?

FlowIT-JIT replied:
Oh, worth mentioning; I took the liberty to move some code from tests/plugins/clipboard/pasteimage.js into tests/plugins/clipboard/_helpers/pasting.js to reduce duplicate code. I needed the exact same code for the new unit test in tests/plugins/clipboard/pasteimagedisabled.js

Comandeer replied:
> Do you still want manual test coverage, or should I just remove it again ?

It's always better to have an additional manual test. For me manual tests are for people and unit ones – for machines.

FlowIT-JIT replied:
Alright, I'll refacter the manual test as you suggested. Thanks 😊

FlowIT-JIT replied:
There. The new manual test (tests/plugins/clipboard/manual/pasteimagedisable.md) is now only about testing the new `clipboard_handleImagePasting` option, as support for image pasting is already covered with tests/plugins/clipboard/manual/pasteimage.md.

Comandeer commented on tests/plugins/clipboard/manual/pasteimagedisable.md:
```diff
@@ -0,0 +1,17 @@
+@bender-ui: collapsed
+@bender-tags: 4.17.0, clipboard, 4874
```
Please, annotate that this test is for a bug:

```suggestion
@bender-tags: 4.17.0, clipboard, 4874, bug
```

Comandeer reviewed (changes requested):
Hi, @FlowIT-JIT, and thanks for your contribution! There are however some issues – please see my inline comments for more info.

Additionally, it will be good to add some unit tests – to check if setting this option to `false` disables built-in handling of pasting images and if the `paste` event is no longer stopped. Could you add such tests?

FlowIT-JIT commented:
@Comandeer Thank you for reviewing my PR. I have made some adjustments based on your feedback.

Comandeer commented on tests/plugins/clipboard/manual/pasteimagedisable.html:
```diff
@@ -0,0 +1,32 @@
+
+<script>
+	( function() {
+		if ( !window.FileReader ) {
```
This test should be ignored in Internet Explorer 10+ as these browsers support File API, but do not support custom MIME types during pasting (only during drag and drop). Probably the most sensible way to ignore this test there would be to do some browser sniffing. Additionally, this test should be ignored on mobile as pasting images is not possible in the same way as on desktops.

So the check should look like this:

```suggestion
		if ( CKEDITOR.env.ie || bender.tools.env.mobile ) {
```

Could you update it?

Comandeer commented on plugins/clipboard/plugin.js:
```diff
@@ -3471,3 +3472,13 @@
 CKEDITOR.config.clipboard_notificationDuration = 10000;
+
+/**
+ * Whether to use clipboard plugin to handle image pasting, turning
```
```suggestion
 * Whether to use clipboard plugin to handle image pasting and dropping, turning
```

Comandeer reviewed (changes requested):
Thx for the updates!

Your changes work correctly, but it seems that some additional tests will be needed, as the new config variable also disables dragging and dropping images. Could you add a manual test and unit test for this case? You can probably reuse some code from `dropimage.js` test. I'm sorry for not noticing it earlier.

It'd be also good to add a unit test for the original case, so checking if some custom paste handler can receive `paste` data after disabling image handling. Could you also add such a test?

FlowIT-JIT commented:
> Thx for the updates!
> 
> Your changes work correctly, but it seems that some additional tests will be needed, as the new config variable also disables dragging and dropping images. Could you add a manual test and unit test for this case? You can probably reuse some code from `dropimage.js` test. I'm sorry for not noticing it earlier.
> 
> It'd be also good to add a unit test for the original case, so checking if some custom paste handler can receive `paste` data after disabling image handling. Could you also add such a test?

Thank you for additional feedback. I should probably have noticed drag and drop support myself - sorry about that. I will look into extending test coverage as suggested.

FlowIT-JIT commented:
I have now added a unit test and manual test to cover image drag and drop.
In addition I extended the unit tests to verify that a custom image handler plugin receives the image when pasting or dropping. I combined this check with the existing unit tests to reduce duplicate code.

Furthermore I renamed the new option from `clipboard_handleImagePasting` to `clipboard_handleImages` as it is not restricted to paste operations alone. In conjunction with this change I also renamed the unit test file from `pasteimagedisabled.js` to `clipboardimagedisabled.js` which execute tests for both paste and drag'n'drop.

Please note that I once again had to do a little refactoring to reduce duplicate code for the new unit test covering image dropping. I moved code from `tests/plugins/clipboard/dropimage.js` to `tests/plugins/clipboard/_helpers/pasting.js`, but this time had to do some minor changes to tear down logic, as outline below:

![image](https://user-images.githubusercontent.com/10479064/134500611-33036244-e413-46e1-8dd5-06280de19fcd.png)

Prior to this change the cleanup was - probably not entirely logical nor correct - performed in setUp(..) as shown below:

![image](https://user-images.githubusercontent.com/10479064/134504849-f453747b-bd2f-41ca-97af-5cb28ad58d22.png)

Comandeer reviewed (approved):
LGTM! I've just updated tests a bit:

* drop test should run in IE10+, so I updated check accordingly,
* we use global `assert` in our tests – I've updated asserts accordingly in your tests,
* I've renamed `assertDropImage` to `assertImageDrop` to achieve better coherency with `assertImagePaste`.

Thanks for the PR!

FlowIT-JIT commented:
@Comandeer Perfect! Thank you 😊
